backend/src/apiserver/server/pipeline_server.go

```diff
@@ -98,427 +108,1027 @@ type PipelineServer struct {
+		return "", util.NewInvalidInputError("Failed to extract pipeline's name as it is too long. Maximum length is %v.", common.MaxFileNameLength)
+	}
+	return pipelineName, nil
+}
 
-	if err := ValidateCreatePipelineRequest(request); err != nil {
-		return nil, err
+// Checks if a user can access a pipeline version.
+func (s *PipelineServer) canAccessPipelineVersion(ctx context.Context, versionId string, resourceAttributes *authorizationv1.ResourceAttributes) error {
+	if !common.IsMultiUserMode() {
+		// Skip authorization if not multi-user mode.
+		return nil
+	}
+	if versionId != "" {
+		namespace, err := s.resourceManager.GetNamespaceFromPipelineVersion(versionId)
+		// Manually allow users to access pipelines in the default namespace.
+		if (namespace == "") || (namespace == model.NoNamespace) || (namespace == s.options.DefaultNamespace) {
+			return nil
+		}
+		if err != nil {
+			return util.Wrap(err, fmt.Sprintf("[PipelineServer %s]: Failed to authorize with the pipeline version id %s.", s.options.ApiVersion, versionId))
+		}
+		resourceAttributes.Namespace = namespace
 	}
+	if resourceAttributes.Namespace == "" {
+		return nil
+	}
+	return s.haveAccess(ctx, resourceAttributes)
+}
 
-	pipelineUrl := request.Pipeline.Url.PipelineUrl
-	resp, err := s.httpClient.Get(pipelineUrl)
-	if err != nil || resp.StatusCode != http.StatusOK {
-		return nil, util.NewInternalServerError(err, "Failed to download the pipeline from %v "+
-			"Please double check the URL is valid and can be accessed by the pipeline system.", pipelineUrl)
+// Checks if a user can access a pipeline.
+func (s *PipelineServer) canAccessPipeline(ctx context.Context, pipelineId string, resourceAttributes *authorizationv1.ResourceAttributes) error {
+	if !common.IsMultiUserMode() {
+		// Skip authorization if not multi-user mode.
+		return nil
 	}
-	pipelineFileName := path.Base(pipelineUrl)
-	pipelineFile, err := ReadPipelineFile(pipelineFileName, resp.Body, MaxFileLength)
-	if err != nil {
-		return nil, util.Wrap(err, "The URL is valid but pipeline system failed to read the file.")
+	if pipelineId != "" {
+		namespace, err := s.resourceManager.GetNamespaceFromPipelineID(pipelineId)
+		// Manually allow users to access pipelines in the default namespace.
+		if (namespace == "") || (namespace == model.NoNamespace) || (namespace == s.options.DefaultNamespace) {
+			return nil
+		}
+		if err != nil {
+			return util.Wrap(err, fmt.Sprintf("[PipelineServer %s]: Failed to authorize with the pipeline id %s.", s.options.ApiVersion, pipelineId))
+		}
+		resourceAttributes.Namespace = namespace
 	}
+	if resourceAttributes.Namespace == "" {
+		return nil
+	}
+	return s.haveAccess(ctx, resourceAttributes)
+}
 
-	pipelineName, err := GetPipelineName(request.Pipeline.Name, pipelineFileName)
+// Checks if a user has access to a resource.
+func (s *PipelineServer) haveAccess(ctx context.Context, resourceAttributes *authorizationv1.ResourceAttributes) error {
```

Comment:
Why do we have both `haveAcess` and `canAccessPipeline`? Do they cover the same function? If not, maybe rename this function for more clarity.

Reply:
I believe this has beed removed. But we have `checkAccess` function that is being called from `canAccessPipelineVersion`, `canAccessPipeline`, and a few more places. This helps to increase code reuse.